lib/optimizer/src/Transformer/ReorderHead.php

```diff
@@ -0,0 +1,372 @@
+     */
+    private function getName(DOMElement $node)
+    {
+        if ($node->hasAttribute(Attribute::CUSTOM_ELEMENT)) {
+            return $node->getAttribute(Attribute::CUSTOM_ELEMENT);
+        }
+
+        if ($node->hasAttribute(Attribute::CUSTOM_TEMPLATE)) {
+            return $node->getAttribute(Attribute::CUSTOM_TEMPLATE);
+        }
+
+        return '';
+    }
+
+    /**
+     * Append all registered nodes to the <head> node.
+     *
+     * @param Document $document Document to append the nodes to.
+     */
+    private function appendToHead(Document $document)
+    {
+        $categories = [
+            'metaCharset',
+            'ampResourceHints',
+            'linkStyleAmpRuntime',
+            'styleAmpRuntime',
+            'metaOther',
+            'scriptAmpRuntime',
+            'scriptAmpViewer',
+            'scriptRenderDelayingExtensions',
+            'scriptNonRenderDelayingExtensions',
+            'linkIcons',
+            'resourceHintLinks', // This should probably be higher, but both the Go and NodeJS optimizers have it here.
+            'linkStylesheetsBeforeAmpCustom',
+            'styleAmpCustom',
+            'others',
+            'styleAmpBoilerplate',
+            'noscript',
+        ];
+
+        foreach ($categories as $category) {
+            if ($this->$category === null) {
+                continue;
+            }
+
+            if ($this->$category instanceof DOMNode) {
+                $node = $document->importNode($this->$category);
+                $document->head->appendChild($node);
+            } elseif (is_array($this->$category)) {
+                // @todo Maybe sort by attribute-name, attribute-value?
+                foreach ($this->$category as $node) {
+                    $node = $document->importNode($node);
+                    $document->head->appendChild($node);
+                }
+            }
+        }
+    }
+
+    /**
+     * Deduplicate and sort custom extensions.
+     */
+    private function deduplicateAndSortCustomNodes()
+    {
+        foreach (['scriptRenderDelayingExtensions', 'scriptNonRenderDelayingExtensions'] as $set) {
+            $sortedNodes = [];
+            foreach ($this->$set as $node) {
+                $sortedNodes[$this->getName($node)] = $node;
+            }
+            ksort($sortedNodes);
+            $this->$set = array_values($sortedNodes);
+        }
+    }
+
+    /**
+     * Check if a given string contains another string, respecting word boundaries..
+     *
+     * @param string $haystack Haystack string to look in.
+     * @param string $needle   Needle string to search for.
+     * @return bool Whether the needle was found in the haystack.
+     */
+    private function containsWord($haystack, $needle)
+    {
+        if (empty($haystack) || empty($needle)) {
+            return false;
+        }
+
+        return preg_match('/\b' . preg_quote($needle, '/') . '\b/i', $haystack);
```

Comment:
This isn't quite right. The `\b` matches a hyphen as a boundary. It should rather be looking for surrounding whitespace and/or start/end line:

```suggestion
        return preg_match('/(^|\s)' . preg_quote($needle, '/') . '(\s|$)/i', $haystack);
```